Hotfix for OpenCG compatibility with 3D geometries

wbinventor opened the pull request:
This PR fixes a few small issues in the OpenCG compatibility module as a result of #188, namely the following:

* Use of `ZCylinder` in place of `Circle` surfaces
* Use of 3D rather than 2D `Lattices`

I've verified that these fixes allow for full 3D modeling of fuel assemblies extracted from the full core 3D OpenCG BEAVRS model (or at least 2D axial "slices" of any assembly anywhere along their axial extent). This should be a quick merge, unless I've missed anything that @samuelshaner might notice.

samuelshaner commented:
It's nice to hear this is working out. I have OpenCG and PWR_benchmarks installed. Do you have a sample input that I could used to test it out?

wbinventor commented:
Sure, below is a script that will run an OpenMC calculation to compute 2-group MGXS and inject them into an OpenMOC calculation for one of the fuel assemblies in the BEAVRS model. I still have a few minor issues to sort out to pefect this data flow (and bounds on z for the OpenMC model are incorrect, but work) but this will work if you want to play around with it. You will need to install the following for this to work:

* **wbinventor/openmc** ("mgxs" branch)
* **wbinventor/OpenMOC** ("materialize-refactor" branch)
* **mit-crpg/OpenCG** ("master" branch)
* **mit-crpg/PWR_benchmarks** ("beavrs_updates" branch)

```python
import openmc
import openmc.mgxs
import openmoc
import opencg
import opencg.plotter
import openmoc.plotter
from openmc.opencg_compatible import get_openmc_geometry
from openmoc.compatible import get_openmoc_geometry
from beavrs.builder import BEAVRS


########   Instantiate an OpenCG BEAVRS model for some fuel assembly  #########

# Instantiate a BEAVRS OpencG model from mit-crpg/PWR_benchmarks
beavrs = BEAVRS(nndc_xs=True)

# Write ALL BEAVRS Materials to a "materials.xml" file
beavrs.write_openmc_materials()

# Get all OpenCG Universes from BEAVRS OpenCG model
all_univ = beavrs.main_universe.get_all_universes()

# Iterate over all Universes to find fuel assembly Lattice
assembly_name = 'Fuel 1.6% enr instr no BAs'
for univ_id, univ in all_univ.items():
    if univ.name == assembly_name:
        fuel_assembly = univ

# Create a Geometry around the fuel assembly
root_cell = opencg.Cell(name='root cell')
root_cell.fill = fuel_assembly

# Make mixed reflective / vacuum boundaries
min_x = opencg.XPlane(x0=root_cell.fill.min_x, boundary='reflective')
max_x = opencg.XPlane(x0=root_cell.fill.max_x, boundary='reflective')
min_y = opencg.YPlane(y0=root_cell.fill.min_y, boundary='reflective')
max_y = opencg.YPlane(y0=root_cell.fill.max_y, boundary='reflective')
min_z = opencg.ZPlane(z0=root_cell.fill.min_z, boundary='reflective')
max_z = opencg.ZPlane(z0=root_cell.fill.max_z, boundary='reflective')

# Add boundaries to the root Cell
root_cell.add_surface(surface=min_x, halfspace=+1)
root_cell.add_surface(surface=max_x, halfspace=-1)
root_cell.add_surface(surface=min_y, halfspace=+1)
root_cell.add_surface(surface=max_y, halfspace=-1)
root_cell.add_surface(surface=min_z, halfspace=+1)
root_cell.add_surface(surface=max_z, halfspace=-1)

# Create a root Universe for this fuel assembly
root_univ = opencg.Universe(universe_id=0, name='root universe')
root_univ.add_cell(root_cell)

# Create an OpenCG Geometry for this fuel assembly
opencg_geometry = opencg.Geometry()
opencg_geometry.root_universe = root_univ


#########   Exporting to OpenMC geometry.xml and materials.xml Files  ##########

# Create an OpenMC Geometry from the OpenCG fuel assembly Geometry
openmc_geometry = get_openmc_geometry(opencg_geometry)
geometry_file = openmc.GeometryFile()
geometry_file.geometry = openmc_geometry
geometry_file.export_to_xml()


##################   Exporting to OpenMC settings.xml File  ###################

settings_file = openmc.SettingsFile()
settings_file.batches = 100
settings_file.inactive = 10
settings_file.particles = 10000
settings_file.statepoint_interval = settings_file.batches
settings_file.ptables = True
settings_file.output = {'tallies': False, 'summary': True}
settings_file.set_source_space('box', opencg_geometry.bounds)
settings_file.export_to_xml()


########################   Build OpenMC MGXS Library  #########################

# Initialize a fine (70-) group MGXS Library for OpenMOC
mgxs_lib = openmc.mgxs.Library(openmc_geometry, by_nuclide=True)
mgxs_lib.energy_groups = openmc.mgxs.EnergyGroups(group_edges=[0,0.625e-6,20.])
mgxs_lib.mgxs_types = ['transport', 'nu-fission', 'nu-scatter matrix', 'chi']
mgxs_lib.domain_type = 'material'
mgxs_lib.build_library()

# Create a "tallies.xml" file for the MGXS Library
tallies_file = openmc.TalliesFile()
mgxs_lib.add_to_tallies_file(tallies_file, merge=True)
tallies_file.export_to_xml()


###############################   Run OpenMC!!  ###############################

executor = openmc.Executor()
executor.run_simulation(mpi_procs=3)


####################   Compute Multi-Group Cross Sections  ####################

# Load the last statepoint and summary files
sp = openmc.StatePoint('statepoint.100.h5')
su = openmc.Summary('summary.h5')
sp.link_with_summary(su)
su.make_opencg_geometry()

# Initialize MGXS Library with OpenMC statepoint data
mgxs_lib.load_from_statepoint(sp)


############################   Build OpenMOC Model  ###########################

# Create an OpenMOC Geometry from the OpenCG Geometry
openmoc_geometry = get_openmoc_geometry(su.opencg_geometry)
openmoc.materialize.load_openmc_mgxs_lib(mgxs_lib, openmoc_geometry)

# Generate tracks
openmoc_geometry.initializeFlatSourceRegions()
track_generator = openmoc.TrackGenerator(openmoc_geometry, 4, 0.1)
track_generator.setZCoord(300.0)
track_generator.generateTracks()


####################   Run OpenMOC Eigenvalue Calculation  ####################

# Run OpenMOC
solver = openmoc.CPUSolver(track_generator)
solver.setNumThreads(4)
solver.setConvergenceThreshold(1E-7)
solver.computeEigenvalue()
solver.printTimerReport()


##################   Use OpenCG and OpenMOC Plotting Tools  ###################

# Use OpenCG to plot the materials
opencg.plotter.plot_materials(su.opencg_geometry, plane='xy', offset=300.)
opencg.plotter.plot_materials(su.opencg_geometry, plane='xz', zlim=[200, 210])
opencg.plotter.plot_materials(su.opencg_geometry, plane='yz', zlim=[200, 210])

# Use OpenMOC to plot the materials
openmoc.plotter.plot_materials(openmoc_geometry, zcoord=300.)
openmoc.plotter.plot_cells(openmoc_geometry, zcoord=300.)
openmoc.plotter.plot_flat_source_regions(openmoc_geometry)

# Use OpenMOC to plot the scalar fluxes
energy_groups = list(range(1, coarse_groups.num_groups+1))
openmoc.plotter.plot_spatial_fluxes(solver, energy_groups=energy_groups)
```

samuelshaner commented on openmoc/compatible/opencg_compatible.py:
```diff
@@ -724,8 +724,8 @@ def get_openmoc_lattice(opencg_lattice):
 
   openmoc_lattice = openmoc.Lattice(lattice_id, name)
   openmoc_lattice.setWidth(width[0], width[1])
```
The z-width should be added as well.

samuelshaner commented:
Line 227 in the `get_openmoc_surface` function where the openmoc `Plane` is initialized should include `C`.

samuelshaner commented:
The `get_opencg_surface` function should also be modified since planes in OpenMOC have `C` as the z-component and `D` as the constant.

samuelshaner commented:
The `get_compatible_opencg_cells` function should also support `x-squareprism` and `y-squareprism` now.

wbinventor commented:
All great comments, thanks for looking through the compatibility module @samuelshaner. I've addressed all of your comments and think we can now support penCG's `x/y/z-squareprism` surface types.

samuelshaner commented on openmoc/compatible/opencg_compatible.py:
```diff
@@ -161,7 +161,8 @@ def get_opencg_surface(openmoc_surface):
     C = openmoc_surface.getC()
-    opencg_surface = opencg.Plane(surface_id, name, boundary, A, B, 0, C)
+    D = openmoc_surface.getD()
+    opencg_surface = opencg.Plane(surface_id, name, boundary, A, B, D, C)
```
I believe C and D should be switched.

wbinventor replied:
Oh wow, embarrassing!

samuelshaner commented on openmoc/compatible/opencg_compatible.py:
```diff
@@ -301,17 +302,34 @@ def get_compatible_opencg_surfaces(opencg_surface):
+    top = opencg.ZPlane(name=name, boundary=boundary, z0=z0+R)
+    surfaces = [left, right, bottom, top]
+
+  elif opencg_surface.type == 'z-squareprism':
```
This should be `x-squareprism` with corresponding y and z surfaces. Also, how do you feel about changing left, right, bottom, and top to ymin, ymax, zmin, and zmax and making this consistent throughout this compatibility module? With the move from strictly 2D in xy to 3D, the bottom and top references can be ambiguous since they could refer to y or z.

wbinventor replied:
Agreed, I like this change and will commit it next.

samuelshaner commented on openmoc/compatible/opencg_compatible.py:
```diff
@@ -387,7 +405,8 @@ def get_compatible_opencg_cells(opencg_cell, opencg_surface, halfspace):
-  if opencg_surface.type == 'z-squareprism':
+  if opencg_surface.type in ['x-squareprism', 'y-squareprism', 
+                             'z-squareprism']:
 
```
Could you modify the comments in lines 411-475 to be ambiguous about the plane directions they are referring to?

samuelshaner replied:
This is more of a personal preference, but you could consider adding some ascii art to describe what the 8 cells correspond to:

    # If Cell is outside the SquarePrism (positive halfspace), add "outside" of Surface halfspaces. 
    # Since OpenMOC does not have a SquarePrism Surface, individual Cells are created 
    # for the 8 Cells that make up the outer region of a SquarePrism.
    #                 |                    |
    #           0     |        1           |    2
    #           ______|____________________|______
    #                 |     SquarePrism    |
    #           7     |   (-)  halfspace   |    3
    #           ______|____________________|______
    #                 |                    |  
    #           6     |        5           |    4
    #                 |                    |

wbinventor replied:
Great, thanks for that, yes this bit of code can be a bit mind-numbing!

wbinventor commented:
@samuelshaner all of your comments have been addressed.

samuelshaner commented on openmoc/compatible/opencg_compatible.py:
```diff
@@ -282,36 +283,81 @@ def get_compatible_opencg_surfaces(opencg_surface):
+  #                 |                    |  
+  #           6     |        5           |    4
+  #                 |                    |
+
```
I was thinking the ascii art would be well placed at line 461 in place of the comment currently on that line.

samuelshaner replied:
Thanks for making the changes @wbinventor! It looks like the comments on lines 439-503 still need to made ambiguous in terms of which surface type they are referring to. With the ascii art reference, I think it could be something as simple as:

    # Subcell 1

samuelshaner commented:
One thought I wanted to bring up is considering modifying `opencg_compatible.py` to be PEP-8 compliant. However, to keep this PR short and sweet, this could be saved for a future PR.

wbinventor commented:
I moved the ASCII art comment as requested. I agree that it should be made PEP-8 compliant, but this can be a future PR since none of our Python modules are PEP-8 compliant at the moment. It should be pretty straightforward - simply tabbing everything with 4 spaces instead of 2, and switching to docstrings comments.

samuelshaner commented on openmoc/compatible/opencg_compatible.py:
```diff
@@ -282,36 +283,67 @@ def get_compatible_opencg_surfaces(opencg_surface):
-    msg = 'Unable to create a compatible OpenMOC Surface from an OpenCG ' \
-          'Surface of type {0} since it already a compatible ' \
-          'Surface type in OpenMOC'.format(opencg_surface.type)
+    msg = 'Unable to create a compatible OpenMC Surface an OpenCG ' \
```
OpenMC -> OpenMOC in two places in this message

samuelshaner commented on openmoc/compatible/opencg_compatible.py:
```diff
@@ -282,36 +283,67 @@ def get_compatible_opencg_surfaces(opencg_surface):
-          'Surface type in OpenMOC'.format(opencg_surface.type)
+    msg = 'Unable to create a compatible OpenMC Surface an OpenCG ' \
+          'Surface of type "{0}" since it already a compatible ' \
+          'Surface type in OpenMC'.format(opencg_surface.type)
```
OpenMC -> OpenMOC

samuelshaner commented:
Once the message is fixed, I'll give this a quick test and then I think it is good to go.

wbinventor commented:
Alright, I fixed OpenMC -> OpenMOC in the compatibility module

On Mon, Oct 19, 2015 at 12:19 PM, sam <notifications@github.com> wrote:

> Once the message is fixed, I'll give this a quick test and then I think it
> is good to go.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/213#issuecomment-149265555>.
>



-- 
Will Boyd
Nuclear Science & Engineering
Massachusetts Institute of Technology
Cell: 423.413.8469
wbinventor@gmail.com
